build: fix build on FreeBSD

knz opened the pull request:
On FreeBSD, rlimit values are signed.

Build was failing due to comparison/assignment between mixed
signed and unsigned types. Forced conversion to the appropriate
type to get build to succeed.

Also added missing -lm.

Supersedes #13809.

mqudsi committed:
build: fix build on FreeBSD

On FreeBSD, rlimit values are signed.

Build was failing due to comparison/assignment between mixed
signed and unsigned types. Forced conversion to the appropriate
type to get build to succeed.

Also added missing -lm.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15336)
<!-- Reviewable:end -->

tamird commented:
Reviewed 5 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/cli/start_jemalloc.go, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaUG0N95xtoFFCDMBx:-KiaUG0N95xtoFFCDMBy:b-unv7e9) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/cli/start_jemalloc.go#L24)):*
> ```Go
> // #cgo LDFLAGS: -ljemalloc
> // #cgo linux LDFLAGS: -lrt -lm -lpthread
> // #cgo freebsd LDFLAGS: -lm
> ```

nit: alphabetize with "linux"

---

*[pkg/server/config_unix.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaUUNUABrU5oUkJVfx:-KiaUUNUABrU5oUkJVfy:bicsxny) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/config_unix.go#L28)):*
> ```Go
> )
> 
> type rlimit struct {
> ```

comment plz!

also don't we want ours to be signed, so the range is smaller?

---

*[pkg/server/rlimit_freebsd.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaVgTnqhefYHDo7XC2:-KiaVgTnqhefYHDo7XC3:b-ffh6m0) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/rlimit_freebsd.go#L15)):*
> ```Go
> // permissions and limitations under the License.
> 
> // +build freebsd
> ```

btw, you don't need this build tag (but you will when it also applies to dragonfly) because of the file name.

---

*[pkg/server/rlimit_freebsd.go, line 35 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaWNts0widPTg2jgog:-KiaWNttj6mkcMlDrjLk:b-t62fvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffce2e7daa0f81ca0dcd764a868f38220db99034/pkg/server/rlimit_freebsd.go#L35)):*
> ```Go
> 		return err
> 	}
> 	if rLimit.Cur == -1 {
> ```

can you add some comments explaining this?

---

*[pkg/server/rlimit_unix.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaUkuMiQN2kSbsoM4Q:-KiaUkuMiQN2kSbsoM4R:b-byrhzk) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/rlimit_unix.go#L15)):*
> ```Go
> // permissions and limitations under the License.
> 
> // +build !windows,!freebsd
> ```

because i've already spelunked into this code in x/sys/unix: freebsd and dragonflybsd are the same in this regard, so let's go ahead and include dragonflybsd here

https://github.com/golang/sys/blob/master/unix/ztypes_dragonfly_amd64.go#L52:L55

---

*[pkg/server/rlimit_unix.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaVblcBvKXiAoue4n6:-KiaVblcBvKXiAoue4n7:b-m1hr6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/rlimit_unix.go#L28)):*
> ```Go
> func getrlimit_nofile(limits *rlimit) error {
> 	var rLimit unix.Rlimit
> 	if err := unix.Getrlimit(unix.RLIMIT_NOFILE, &rLimit); err != nil {
> ```

hm, isn't `rlimit` convertible to `unix.Rlimit` in this case? I'm wondering if you can just:

```
return unix.Getrlimit(unix.RLIMIT_NOFILE, (*unix.Rlimit)(&rlimit))
```
or a similar construction.

The same comment applies to the setter above.

---

*[pkg/server/status/runtime_jemalloc.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaUMXO2wmz4hrAgS1F:-KiaUMXO2wmz4hrAgS1G:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/status/runtime_jemalloc.go#L30)):*
> ```Go
> // #cgo darwin LDFLAGS: -u_je_zone_register
> // #cgo linux LDFLAGS: -lrt -lm -lpthread
> // #cgo freebsd LDFLAGS: -lm
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/cli/start_jemalloc.go, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaUG0N95xtoFFCDMBx:-Kiac9P_qp6KKMvNPyui:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/cli/start_jemalloc.go#L24)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: alphabetize with "linux"
</blockquote></details>

Done.

---

*[pkg/server/config_unix.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaUUNUABrU5oUkJVfx:-KiacaHK8kl5TP0ZBD3W:brt48c) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/config_unix.go#L28)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

comment plz!

also don't we want ours to be signed, so the range is smaller?
</blockquote></details>

Added a comment.
I don't like a signed type here because it suggests that a negative value is valid, which would be a wrong idea.

---

*[pkg/server/rlimit_freebsd.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaVgTnqhefYHDo7XC2:-KiacmDLk3lDPNl71SkE:b-m1s000) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/rlimit_freebsd.go#L15)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

btw, you don't need this build tag (but you will when it also applies to dragonfly) because of the file name.
</blockquote></details>

I added dragonfly right away.

---

*[pkg/server/rlimit_freebsd.go, line 35 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaWNts0widPTg2jgog:-KiacpdKhYAG9cgeeUDu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffce2e7daa0f81ca0dcd764a868f38220db99034/pkg/server/rlimit_freebsd.go#L35)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can you add some comments explaining this?
</blockquote></details>

Done.

---

*[pkg/server/rlimit_unix.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaUkuMiQN2kSbsoM4Q:-Kiaczn22BJBtKH4Qlmz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/rlimit_unix.go#L15)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

because i've already spelunked into this code in x/sys/unix: freebsd and dragonflybsd are the same in this regard, so let's go ahead and include dragonflybsd here

https://github.com/golang/sys/blob/master/unix/ztypes_dragonfly_amd64.go#L52:L55
</blockquote></details>

Done.

---

*[pkg/server/rlimit_unix.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaVblcBvKXiAoue4n6:-Kiae2tUS1bMOvVKqDZh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/rlimit_unix.go#L28)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

hm, isn't `rlimit` convertible to `unix.Rlimit` in this case? I'm wondering if you can just:

```
return unix.Getrlimit(unix.RLIMIT_NOFILE, (*unix.Rlimit)(&rlimit))
```
or a similar construction.

The same comment applies to the setter above.
</blockquote></details>

Done.

---

*[pkg/server/status/runtime_jemalloc.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaUMXO2wmz4hrAgS1F:-Kiae_KpiXcg1xX6Rm05:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/status/runtime_jemalloc.go#L30)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/cli/start_jemalloc.go, line 23 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiagyMg7sh6pHKtIDvb:-KiagyMh94C0K0my_EYw:bw08tpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b501315499cf8bdf0bab95a14ad9f0128f76e8d4/pkg/cli/start_jemalloc.go#L23)):*
> ```Go
> // #cgo CPPFLAGS: -DJEMALLOC_NO_DEMANGLE
> // #cgo LDFLAGS: -ljemalloc
> // #cgo dragonfly LDFLAGS: -lm
> ```

nit: this can be `#cgo dragonfly freebsd LDFLAGS`

---

*[pkg/server/config_unix.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaUUNUABrU5oUkJVfx:-KiahTUwCB-0DuRd5rQV:b-cw7vl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/config_unix.go#L28)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Added a comment.
I don't like a signed type here because it suggests that a negative value is valid, which would be a wrong idea.
</blockquote></details>

Sounds good. I think the abstracting type should stay here, explained more elsewhere.

---

*[pkg/server/rlimit_freebsd.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaVgTnqhefYHDo7XC2:-Kiah8Gp12hzLlyaw8eJ:b1w1wo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/rlimit_freebsd.go#L15)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I added dragonfly right away.
</blockquote></details>

cool. can you rename the file?

---

*[pkg/server/rlimit_freebsd.go, line 25 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiahHbxhvpfoe4hOUXl:-KiahHbxhvpfoe4hOUXm:bjwmb7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/b501315499cf8bdf0bab95a14ad9f0128f76e8d4/pkg/server/rlimit_freebsd.go#L25)):*
> ```Go
> )
> 
> // rlimit is a replacement struct for `unix.Rlimit` which abstracts
> ```

this definition can be in the common (not platform specific) file. if it happens to be equivalent to `unix.Rlimit` it will still be assignable.

---

*[pkg/server/rlimit_freebsd.go, line 42 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-Kiah6ibTqt_BFei9xFM:-Kiah6ibTqt_BFei9xFN:bkywvs6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b501315499cf8bdf0bab95a14ad9f0128f76e8d4/pkg/server/rlimit_freebsd.go#L42)):*
> ```Go
> 		return err
> 	}
> 	// Some (legacy?) FreeBSD platforms had RLIMIT_INFINITY set to -1.
> ```

dragonfly too, right?

---

*[pkg/server/status/runtime_jemalloc.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaUMXO2wmz4hrAgS1F:-KiahRZDd8_UyMZnVWvj:b-cabdg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/status/runtime_jemalloc.go#L30)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

one line?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/cli/start_jemalloc.go, line 23 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiagyMg7sh6pHKtIDvb:-Kib4HV2R11FGemOaaIZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b501315499cf8bdf0bab95a14ad9f0128f76e8d4/pkg/cli/start_jemalloc.go#L23)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: this can be `#cgo dragonfly freebsd LDFLAGS`
</blockquote></details>

Done.

---

*[pkg/server/config_unix.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaUUNUABrU5oUkJVfx:-Kib4sJa8mdd1h589qGg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/config_unix.go#L28)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Sounds good. I think the abstracting type should stay here, explained more elsewhere.
</blockquote></details>

Done.

---

*[pkg/server/rlimit_freebsd.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaVgTnqhefYHDo7XC2:-Kib4OiSxkthW8rYpipQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/rlimit_freebsd.go#L15)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

cool. can you rename the file?
</blockquote></details>

Done.

---

*[pkg/server/rlimit_freebsd.go, line 25 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiahHbxhvpfoe4hOUXl:-Kib4tI6840VJC_RBsFl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b501315499cf8bdf0bab95a14ad9f0128f76e8d4/pkg/server/rlimit_freebsd.go#L25)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this definition can be in the common (not platform specific) file. if it happens to be equivalent to `unix.Rlimit` it will still be assignable.
</blockquote></details>

Done.

---

*[pkg/server/rlimit_freebsd.go, line 42 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-Kiah6ibTqt_BFei9xFM:-Kib4QFMYgV2j5TDddhK:b5maw25) ([raw file](https://github.com/cockroachdb/cockroach/blob/b501315499cf8bdf0bab95a14ad9f0128f76e8d4/pkg/server/rlimit_freebsd.go#L42)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

dragonfly too, right?
</blockquote></details>

Dragonfly *is* FreeBSD when it comes to the kernel and everything related to syscalls.

---

*[pkg/server/status/runtime_jemalloc.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-KiaUMXO2wmz4hrAgS1F:-Kib4Z3wZlQKgY9_I5s4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36eb587aee31535e6571092650a58d3c3045b964/pkg/server/status/runtime_jemalloc.go#L30)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

one line?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Thank you!

---

Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336#-:-KibCevoIeY4sleTPRda:b-vdwr1z)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 6 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15336)*
<!-- Sent from Reviewable.io -->

knz commented:
Feel free to merge when tests pass. I'm afk until tomorrow.

tamird commented:
Will do.

On Tue, Apr 25, 2017 at 6:51 PM, kena <notifications@github.com> wrote:

> Feel free to merge when tests pass. I'm afk until tomorrow.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/15336#issuecomment-297187984>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPGrUS6L4taVHpmr4_E41lH2ubjYPks5rznj_gaJpZM4NH6rp>
> .
>
